extensions/amp-bind/0.1/bind-expression.js

```diff
@@ -22,25 +23,232 @@ import {parser} from './bind-expr-impl';
 export class BindExpression {
   /**
    * @param {string} expressionString
+   * @throws {Error} On malformed expressions.
    */
   constructor(expressionString) {
     /** @const {string} */
     this.expressionString = expressionString;
+
+    /** {!./bind-expr-defines.ASTNode} */
+    this.ast_ = parser.parse(this.expressionString);
+
+    /** @const {!Object<string, !Object<string, Function>>} */
+    this.functionWhitelist_ = this.createFunctionWhitelist_();
   }
 
   /**
    * Evaluates the expression given a scope.
    * @param {!Object} scope
+   * @throws {Error} On illegal function invocation.
    * @return {*}
    */
   evaluate(scope) {
-    // TODO(choumx): Improve performance by extracting AST construction
-    // and generating evaluator with Function constructor.
-    try {
-      parser.yy = scope;
-      return parser.parse(this.expressionString);
-    } finally {
-      parser.yy = null;
+    return this.eval_(this.ast_, scope);
+  }
+
+  /**
+   * Recursively evaluates and returns value of `node` and its children.
+   * @param {?./bind-expr-defines.ASTNode} node
+   * @param {!Object} scope
+   * @throws {Error}
+   * @return {*}
+   * @private
+   */
+  eval_(node, scope) {
+    if (!node) {
+      return null;
+    }
+
+    const {type, args, value} = node;
+
+    if (type === ASTNodeType.LITERAL) {
+      return value;
+    }
+
+    // Concise helper function for evaluating child nodes.
+    const e = node => this.eval_(node, scope);
+
+    switch (type) {
+      case ASTNodeType.EXPRESSION:
+        return e(args[0]);
+
+      case ASTNodeType.INVOCATION:
```

Comment:
This won't fly because it is prohibited by our CSP and I absolutely positively do not want to open it up.